src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -138,54 +190,61 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
                     }
                 })
 
-                Result.Ok(description)
+                UpdateResult.Ok(description)
             } catch (e: ExecutionException) {
-                Result.Err(e)
+                UpdateResult.Err(e)
             }
         }
 
         override fun onSuccess() {
-            val result = requireNotNull(result)
+            val r = requireNotNull(result)
 
-            when (result) {
-                is Result.Err -> {
-                    showBalloon("Project '${module.project.name}' update failed: ${result.error.message}", MessageType.ERROR)
-                    LOG.info("Project '${module.project.name}' update failed", result.error)
-                }
-                is Result.Ok  ->
-                    ApplicationManager.getApplication().runWriteAction {
-                        if (!module.isDisposed) {
-                            val libraryRoots =
-                                result.cargoProject.packages
-                                    .filter { !it.isModule }
-                                    .mapNotNull { it.virtualFile }
+            lock (lock) {
+                updateModuleDependencies(r);
```

Comment:
Is this intentionally not called in `setState`? If it is not the case then it makes sense to extract `lock, updateLibs, notify listeners` piece from `onSuccess` and `setState`.